extra_data/write_cxi.py

```diff
@@ -250,10 +335,48 @@ def write(self, filename, fillvalues=None):
                 )
                 mask.attrs['axes'] = axes_s
 
-            dgrp['experiment_identifier'] = h5py.SoftLink('/entry_1/experiment_identifier')
+            dgrp['experiment_identifier'] = h5py.SoftLink(
+                '/entry_1/experiment_identifier')
 
-            f['entry_1/data_1'] = h5py.SoftLink('/entry_1/instrument_1/detector_1')
+            f['entry_1/data_1'] = h5py.SoftLink(
+                '/entry_1/instrument_1/detector_1')
 
             dgrp.create_dataset('module_identifier', data=self.modulenos)
 
         log.info("Finished writing virtual CXI file")
+
+
+class JungfrauCXIWriter(VirtualCXIWriter):
+    """
+    Machinery to write VDS files for Jungfrau data in the same format
+    as AGIPD/LPD virtual datasets.
+
+    Parameters
+    ----------
+
+    detdata: extra_data.components.JUNGFRAU
+      The detector data interface for the data to gather in this file.
+    """
+    def __init__(self, detdata) -> None:
+        # Check number of cells
+        src = next(iter(detdata.source_to_modno))
+        h5file = detdata.data._source_index[src][0].file
+        image = h5file['INSTRUMENT'][src][
+            detdata._main_data_key.replace('.', '/')]
+        self._n_cells = image.shape[1]
```

Comment:
There's now a neater way of getting the shape of one entry:

```python
keydata = detdata.data[src, 'data.adc']
n_cells = keydata.entry_shape[0]
```

[entry_shape](https://extra-data.readthedocs.io/en/latest/reading_files.html#extra_data.KeyData.entry_shape) in the docs

Reply:
Thank you, this looks much better!